
  <h1 class="display-3">Insights on User Behaviour of using Game App in Python</h1>
  <h3>For Churn Prediction</h3>
  <hr>

In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
from pandas import Series, DataFrame

import sklearn
from datetime import datetime

<a id="ImportingDatasets"></a>
# Importing Dataset
<hr>

In [2]:
# Importing Datset
gameuser_data = pd.read_csv("F:\Lectures\Data Science\Projects\Python\Datasets\data_vis_test.csv")
gameuser_data = pd.DataFrame(gameuser_data) # Creating a pandas data frame and storing it in a vector
gameuser_data.shape # Shape gives you total number of observations and variables present in datasets

(10000, 4)

In [3]:
gameuser_data.info() # Info gives you name of each variable with the data type associated with it.
gameuser_data.head(10) # Head gives you first few rows of the dataset. (3 is the nummber of rows)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 4 columns):
user_id                        9998 non-null object
first_open_timestamp_micros    9995 non-null float64
mobile_brand_name              9383 non-null object
device_model                   10000 non-null object
dtypes: float64(1), object(3)
memory usage: 312.6+ KB


user_id  first_open_timestamp_micros mobile_brand_name  \
0  59742df346ab8ae80d7320e5                 1.509260e+15           Samsung   
1  59d23b4cb2cb0511457c3718                 1.506950e+15            Xiaomi   
2  59c8671a625fbbd425d735ae                 1.506310e+15               NaN   
3  591ff88228d8f8c86bfd75bf                 1.506840e+15              OPPO   
4  59feaccfa3c446c82ca3e68c                 1.509860e+15              Sony   
5  59f13874bb299b692aeef502                 1.508980e+15           Samsung   
6  59d4f647bb299b692add0d24                 1.506860e+15               NaN   
7  59d8bafdfc67610b4500232b                 1.506880e+15            Xiaomi   
8  59cc674efc67610b45f1c83b                 1.506570e+15            Xiaomi   
9  58b072b134182301374ff77d                 1.506420e+15              Vivo   

    device_model  
0       SM-J200G  
1       Redmi 4A  
2  Swipe_Elite_3  
3           A37f  
4          C6902  
5       SM-J700F  
6      OctaOne_5  
7   Redmi Note 4  
8   Redmi Note 4  
9      vivo Y51L

#### From above it can be seen that the given dataset has 10K observations and 4 variables. 

<a id="FindNULLVal"></a>
# Converting Timestamp into Date and Time
<hr>

#### Assuming the Timestamps are as per the Indian standard time.

In [4]:
gameuser_data['first_open_timestamp_micros'] = pd.to_datetime(gameuser_data['first_open_timestamp_micros'], unit='us')

In [5]:
gameuser_data.head(3)

user_id first_open_timestamp_micros mobile_brand_name  \
0  59742df346ab8ae80d7320e5         2017-10-29 06:53:20           Samsung   
1  59d23b4cb2cb0511457c3718         2017-10-02 13:13:20            Xiaomi   
2  59c8671a625fbbd425d735ae         2017-09-25 03:26:40               NaN   

    device_model  
0       SM-J200G  
1       Redmi 4A  
2  Swipe_Elite_3

#### Sperating 

In [6]:
gameuser_data["Date"] = gameuser_data['first_open_timestamp_micros'].dt.date
gameuser_data["Time"] = gameuser_data['first_open_timestamp_micros'].dt.time

In [7]:
gameuser_data.head(10)

user_id first_open_timestamp_micros mobile_brand_name  \
0  59742df346ab8ae80d7320e5         2017-10-29 06:53:20           Samsung   
1  59d23b4cb2cb0511457c3718         2017-10-02 13:13:20            Xiaomi   
2  59c8671a625fbbd425d735ae         2017-09-25 03:26:40               NaN   
3  591ff88228d8f8c86bfd75bf         2017-10-01 06:40:00              OPPO   
4  59feaccfa3c446c82ca3e68c         2017-11-05 05:33:20              Sony   
5  59f13874bb299b692aeef502         2017-10-26 01:06:40           Samsung   
6  59d4f647bb299b692add0d24         2017-10-01 12:13:20               NaN   
7  59d8bafdfc67610b4500232b         2017-10-01 17:46:40            Xiaomi   
8  59cc674efc67610b45f1c83b         2017-09-28 03:40:00            Xiaomi   
9  58b072b134182301374ff77d         2017-09-26 10:00:00              Vivo   

    device_model        Date      Time  
0       SM-J200G  2017-10-29  06:53:20  
1       Redmi 4A  2017-10-02  13:13:20  
2  Swipe_Elite_3  2017-09-25  03:26:40  
3           A37f  2017-10-01  06:40:00  
4          C6902  2017-11-05  05:33:20  
5       SM-J700F  2017-10-26  01:06:40  
6      OctaOne_5  2017-10-01  12:13:20  
7   Redmi Note 4  2017-10-01  17:46:40  
8   Redmi Note 4  2017-09-28  03:40:00  
9      vivo Y51L  2017-09-26  10:00:00

<a id="FindNULLVal"></a>
# Finding and Treating NULL Values
<hr>

In [8]:
gameuser_data.isnull().sum() # Gives Variable wise NaN values

user_id                          2
first_open_timestamp_micros      5
mobile_brand_name              617
device_model                     0
Date                             5
Time                             5
dtype: int64

* As, you can see the variables **'user_id'** and **'first_open_timestamp_micros'** have in total 7 NaN values. Since we can not generate or define these values as user input, it is better to remove the observations associated with it.
* **Note:** The NaN values of **Date** and **Time** variable will be removed with the NaN values of **first_open_timestamp_micros** variable.

In [9]:
gameuser_data = gameuser_data[pd.notnull(gameuser_data['user_id'])]
gameuser_data = gameuser_data[pd.notnull(gameuser_data['first_open_timestamp_micros'])]

In [10]:
gameuser_data.isnull().sum()

user_id                          0
first_open_timestamp_micros      0
mobile_brand_name              617
device_model                     0
Date                             0
Time                             0
dtype: int64

In [11]:
gameuser_data.shape

(9993, 6)

* Now, it can be seen that there are still 617 null values in **'mobile_brand_name'** variable, which is **6.17%** data of total data. And, it is not  a good idea to delete this much of data. So, its better to find a way to fill these NaN values in  **'mobile_brand_name'** variable. <br> 
* So, what we are going to do is : we will replace the NaN values in **'mobile_brand_name'** according to first few letters of **'device_model' variable**.<br>
* **Eg:** Wherever there will be **'SM'** in **'mobile_brand_name'** it wil be replaced with **'Samsung'**

In [12]:
gameuser_data.loc[gameuser_data['device_model'].astype(str).str[0:2] == 'SM', 'mobile_brand_name'] = 'Samsung'
gameuser_data.loc[gameuser_data['device_model'].astype(str).str[0:5] == 'Redmi', 'mobile_brand_name'] = 'Xiaomi'
gameuser_data.loc[gameuser_data['device_model'].astype(str).str[0:2] == 'MI', 'mobile_brand_name'] = 'Xiaomi'
gameuser_data.loc[gameuser_data['device_model'].astype(str).str[0:4] == 'vivo', 'mobile_brand_name'] = 'Vivo'

gameuser_data.loc[gameuser_data['device_model'].astype(str).str[0:4] == 'Moto', 'mobile_brand_name'] = 'Motorola'
gameuser_data.loc[gameuser_data['device_model'].astype(str).str[0:8] == 'Micromax', 'mobile_brand_name'] = 'Micromax'
gameuser_data.loc[gameuser_data['device_model'].astype(str).str[0:2] == 'LS', 'mobile_brand_name'] = 'LYF'
gameuser_data.loc[gameuser_data['device_model'].astype(str).str[0:6] == 'Lenovo', 'mobile_brand_name'] = 'Lenovo'
gameuser_data.loc[gameuser_data['device_model'].astype(str).str[0:4] == 'itel', 'mobile_brand_name'] = 'itel'
gameuser_data.loc[gameuser_data['device_model'].astype(str).str[0:9] == 'Panasonic', 'mobile_brand_name'] = 'Panasonic'
gameuser_data.loc[gameuser_data['device_model'].astype(str).str[0:5] == 'ELUGA', 'mobile_brand_name'] = 'Panasonic'
gameuser_data.loc[gameuser_data['device_model'].astype(str).str[0:2] == 'TA', 'mobile_brand_name'] = 'Nokia'

gameuser_data.loc[gameuser_data['device_model'].astype(str).str[0:3] == 'X28', 'mobile_brand_name'] = 'Lava'
gameuser_data.loc[gameuser_data['device_model'].astype(str).str[0:3] == 'X41', 'mobile_brand_name'] = 'Lava'
gameuser_data.loc[gameuser_data['device_model'].astype(str).str[0:3] == 'X50', 'mobile_brand_name'] = 'Lava'
gameuser_data.loc[gameuser_data['device_model'].astype(str).str[0:3] == 'Z60', 'mobile_brand_name'] = 'Lava'
gameuser_data.loc[gameuser_data['device_model'].astype(str).str[0:3] == 'Z90', 'mobile_brand_name'] = 'Lava'
gameuser_data.loc[gameuser_data['device_model'].astype(str).str[0:3] == 'A97', 'mobile_brand_name'] = 'Lava'
gameuser_data.loc[gameuser_data['device_model'].astype(str).str[0:4] == 'iris', 'mobile_brand_name'] = 'Lava'

gameuser_data.loc[gameuser_data['device_model'].astype(str).str[0:5] == 'Elite', 'mobile_brand_name'] = 'Elite'
gameuser_data.loc[gameuser_data['device_model'].astype(str).str[0:3] == 'Era', 'mobile_brand_name'] = 'Xolo'
gameuser_data.loc[gameuser_data['device_model'].astype(str).str[0:5] == 'INTEX', 'mobile_brand_name'] = 'Intex'

In [13]:
gameuser_data.isnull().sum()

user_id                          0
first_open_timestamp_micros      0
mobile_brand_name              149
device_model                     0
Date                             0
Time                             0
dtype: int64

* As you can see we have 149 NaN values which is 1.5% of the entire data. We can drop these rows now.

In [14]:
gameuser_data = gameuser_data[pd.notnull(gameuser_data['mobile_brand_name'])]

In [15]:
gameuser_data.isnull().sum()

user_id                        0
first_open_timestamp_micros    0
mobile_brand_name              0
device_model                   0
Date                           0
Time                           0
dtype: int64

In [16]:
gameuser_data.shape

(9844, 6)

In [17]:
gameuser_data.to_csv('F:\Lectures\Data Science\Projects\Python\Datasets\gameuser_data_updated.csv',index=False,encoding='utf-8') 

<a id="CountDSLevels"></a>
# Counting Levels in Datasets
<hr>

Now we have cleaned data. And we can now count the levels of variables in dataset, which is important because it shows the distribution of elements in variable. In addation to that it also shows the wrongly entred elements.

In [37]:
pd.options.display.max_rows
gameuser_data.mobile_brand_name.value_counts()

Samsung         3079
Xiaomi          1527
Vivo            1061
OPPO             929
Lenovo           564
Motorola         490
Micromax         363
LYF              276
Gionee           245
Lava             150
Intex             93
HTC               91
Coolpad           88
Panasonic         88
Asus              85
OnePlus           81
Huawei            75
Sony              64
LeTV              47
itel              43
LG                42
Karbonn           41
YU                41
Xolo              31
Nokia             29
LeEco             24
iZoom             23
InFocus           18
Google            16
Videocon          13
iBall             12
Elite             10
Alcatel            7
Ziox               7
Tecno              6
Spice              6
SonyEricsson       5
Ampe               5
Allview            5
Celkon             5
Acer               5
Magicon            4
Meitu              4
BlackBerry         4
Lephone            3
Sansui             3
Infinix            3
Swipe        

In [19]:
pd.options.display.max_rows
gameuser_data.device_model.value_counts()

SM-J200G                   550
Redmi Note 4               462
SM-J700F                   405
SM-J210F                   309
A37f                       275
Redmi 3S                   253
Redmi 4A                   252
SM-G610F                   242
Redmi Note 3               218
Redmi 4                    199
vivo 1606                  193
A1601                      160
vivo Y21L                  154
A33f                       147
vivo 1610                  145
vivo Y51L                  140
SM-J710F                   138
SM-J500F                   118
A37fw                      118
SM-J701F                   104
CPH1701                    104
vivo 1603                   98
vivo 1601                   86
SM-G570F                    79
SM-G615F                    75
Moto G (4)                  72
SM-G550FY                   70
SM-G600FY                   66
SM-G531F                    62
CPH1609                     61
                          ... 
Astra Colors 4G              1
HTC One 

<a id="Plots"></a>
# Plot and Graphs 
<hr>

Now, we are going to make some graphs to draw some insights for our cleaned dataset. <br>
**Note:** I tried plotting graphs with Matplotlib and Seaborn. But the graphs was looking messy and taking a long time to generate. So, I will be using Tableau instead to plot the graphs. <br>
**Note:** To view the below graphs you might have to sign up and login to Tableau Online account. [[Sign Up]](https://www.tableau.com/products/cloud-bi?ab=1#form) [[Sign In]](https://sso.online.tableau.com/public/idp/SSO)

In [32]:
%%HTML
<script type='text/javascript' src='https://dub01.online.tableau.com/javascripts/api/viz_v1.js'></script><div class='tableauPlaceholder' style='width: 1366px; height: 582px;'><object class='tableauViz' width='1366' height='582' style='display:none;'><param name='host_url' value='https%3A%2F%2Fdub01.online.tableau.com%2F' /> <param name='embed_code_version' value='3' /> <param name='site_root' value='&#47;t&#47;dsatyajit' /><param name='name' value='Absentia_Intern_Test&#47;AppOccurences' /><param name='tabs' value='no' /><param name='toolbar' value='no' /><param name='showAppBanner' value='false' /><param name='filter' value='iframeSizedToWindow=true' /></object></div>

The above bar graph shows the number of time the game app was opened by users according to the mobile brand. As you can see the users of mobile brands like **Samsung, Xiaomi, Vivo, OPPO, Lenovo, Motorola, Micromax, LYF, Gionee and Lava** are the top 10 mobile brands users who have opened the game app most of the time and these top 10 brands make about **88.21%** of total data.

In [24]:
%%HTML
<script type='text/javascript' src='https://dub01.online.tableau.com/javascripts/api/viz_v1.js'></script><div class='tableauPlaceholder' style='width: 1366px; height: 582px;'><object class='tableauViz' width='1366' height='582' style='display:none;'><param name='host_url' value='https%3A%2F%2Fdub01.online.tableau.com%2F' /> <param name='embed_code_version' value='3' /> <param name='site_root' value='&#47;t&#47;dsatyajit' /><param name='name' value='Absentia_Intern_Test&#47;AppOccurences-ModelWise' /><param name='tabs' value='no' /><param name='toolbar' value='no' /><param name='showAppBanner' value='false' /><param name='filter' value='iframeSizedToWindow=true' /></object></div>

The above tree graph shows the number of time the game app was opened by the user according to **mobile model** and separated by **mobile Brands**. If you hover over each block of the graph it will show you the total percentage share and number of occurrences of that particular mobile model in the whole population.

In [25]:
%%HTML
<script type='text/javascript' src='https://dub01.online.tableau.com/javascripts/api/viz_v1.js'></script><div class='tableauPlaceholder' style='width: 1366px; height: 582px;'><object class='tableauViz' width='1366' height='582' style='display:none;'><param name='host_url' value='https%3A%2F%2Fdub01.online.tableau.com%2F' /> <param name='embed_code_version' value='3' /> <param name='site_root' value='&#47;t&#47;dsatyajit' /><param name='name' value='Absentia_Intern_Test&#47;PercentWiseMobModel' /><param name='tabs' value='no' /><param name='toolbar' value='no' /><param name='showAppBanner' value='false' /><param name='filter' value='iframeSizedToWindow=true' /></object></div>

The above tree graph shows the percentage wise mobile models who opened the game app. It can be seen that though mobile brand wise **Samsung** dominates the game opening app but when it comes to mobile model wise comparison the other brands like **Xiaomi, Vivo and OPPO** have a big chunk of game openings associated with them too. <br>
**So, if you look at first 24 blocks in this graph (vertically selected) which covers 50.27% of data which shows the other mobiles brands have mobile models close to Samsung.** Shown in Table below.


In [38]:
%%HTML
<script type='text/javascript' src='https://dub01.online.tableau.com/javascripts/api/viz_v1.js'></script><div class='tableauPlaceholder' style='width: 1366px; height: 582px;'><object class='tableauViz' width='1366' height='582' style='display:none;'><param name='host_url' value='https%3A%2F%2Fdub01.online.tableau.com%2F' /> <param name='embed_code_version' value='3' /> <param name='site_root' value='&#47;t&#47;dsatyajit' /><param name='name' value='Absentia_Intern_Test&#47;PercentMobModels' /><param name='tabs' value='no' /><param name='toolbar' value='no' /><param name='showAppBanner' value='false' /><param name='filter' value='iframeSizedToWindow=true' /></object></div>

In [27]:
%%HTML
<script type='text/javascript' src='https://dub01.online.tableau.com/javascripts/api/viz_v1.js'></script><div class='tableauPlaceholder' style='width: 1366px; height: 582px;'><object class='tableauViz' width='1366' height='582' style='display:none;'><param name='host_url' value='https%3A%2F%2Fdub01.online.tableau.com%2F' /> <param name='embed_code_version' value='3' /> <param name='site_root' value='&#47;t&#47;dsatyajit' /><param name='name' value='Absentia_Intern_Test&#47;AppHours-ModelWise' /><param name='tabs' value='no' /><param name='toolbar' value='no' /><param name='showAppBanner' value='false' /><param name='filter' value='iframeSizedToWindow=true' /></object></div>

The above tree graph is seperated by mobile models which shows total number of hours the game app was open by each mobile model. So, if we compare by how much time the game app was opened and by which mobile model. We can see that more than 50% of time the game app was opened by mobile brands like **Samsung, Xiaomi, Vivo, OPPO, Lenovo, Motorola, Micromax, LYF, Gionee and Lava**.

In [28]:
%%HTML
<script type='text/javascript' src='https://dub01.online.tableau.com/javascripts/api/viz_v1.js'></script><div class='tableauPlaceholder' style='width: 1366px; height: 582px;'><object class='tableauViz' width='1366' height='582' style='display:none;'><param name='host_url' value='https%3A%2F%2Fdub01.online.tableau.com%2F' /> <param name='embed_code_version' value='3' /> <param name='site_root' value='&#47;t&#47;dsatyajit' /><param name='name' value='Absentia_Intern_Test&#47;QuarterlyOpened' /><param name='tabs' value='no' /><param name='toolbar' value='no' /><param name='showAppBanner' value='false' /><param name='filter' value='iframeSizedToWindow=true' /></object></div>

From this table, it can be seen that in 2017 the **game app is opened more in the fourth quarter of 2017** as compared with the third quarter. The app access **was increased by 97.94%** in Q4 of 2017. (See the grand total at bottom of the table)

In [29]:
%%HTML
<script type='text/javascript' src='https://dub01.online.tableau.com/javascripts/api/viz_v1.js'></script><div class='tableauPlaceholder' style='width: 1366px; height: 582px;'><object class='tableauViz' width='1366' height='582' style='display:none;'><param name='host_url' value='https%3A%2F%2Fdub01.online.tableau.com%2F' /> <param name='embed_code_version' value='3' /> <param name='site_root' value='&#47;t&#47;dsatyajit' /><param name='name' value='Absentia_Intern_Test&#47;WeekdayOpened' /><param name='tabs' value='no' /><param name='toolbar' value='no' /><param name='showAppBanner' value='false' /><param name='filter' value='iframeSizedToWindow=true' /></object></div>

**Assuming that the given data of users is based on India data:** <br>
The above graph shows the number of time the game opened on all days of the week in last two quarters of 2017. As you can see in Q3 only on Fridays and Saturdays there was high volume app openings as compared to Q4. But in Q4, you can see a high increase of app opening on Sundays and Mondays.  Also, there is the high increase of app openings in all the weekdays of Q4 as compared to all the weekdays of Q3. <br>
This could happen because in India there are more holidays in Q4 than Q3. Like Diwali and Christmas Holidays comes in Q4. <br>
_**Note:** The percentage shown in graph is the percent share of app opened each quarter._

In [30]:
%%HTML
<script type='text/javascript' src='https://dub01.online.tableau.com/javascripts/api/viz_v1.js'></script><div class='tableauPlaceholder' style='width: 1366px; height: 582px;'><object class='tableauViz' width='1366' height='582' style='display:none;'><param name='host_url' value='https%3A%2F%2Fdub01.online.tableau.com%2F' /> <param name='embed_code_version' value='3' /> <param name='site_root' value='&#47;t&#47;dsatyajit' /><param name='name' value='Absentia_Intern_Test&#47;BrandWiseOpenedWeekdays' /><param name='tabs' value='no' /><param name='toolbar' value='no' /><param name='showAppBanner' value='false' /><param name='filter' value='iframeSizedToWindow=true' /></object></div>

This graph shows app opened on weekdays of top 15 mobile brands of two quarters of 2017. Here you can see that there is a high volume of app opening from Thursdays to Sundays in both the quarters. But in Q4 on Mondays, there is also a high volume of openings. 

<a id="Plots"></a>
# Conclusion
<hr>

The game app opening is dominated by mobile users of brands like **Samsung, Xiaomi, Vivo, OPPO, Lenovo, Motorola, Micromax, LYF, Gionee and Lava** out of 69 mobile brands. And these mobile brand makes 88.21% of the total data. It is seen that though mobile brand wise Samsung dominates the game opening app but when it comes to mobile model wise comparison the other brands like **Xiaomi, Vivo and OPPO** have a big chunk of game openings associated with them too. Mobile models of these brands make 50.27% of data. <br>
In addition to this, the users of above top 10 mobile brands dominate the number of hours the game app opening. However, Samsung dominates in both accessing app by time and by the number of occurrences. <br>
Furthermore, it is seen that the game app opening was **increased by  97.94% in Q4 of 2017**, as compared to Q3. This could happen because in India there are more holidays in Q4 than Q3. Like Diwali and Christmas Holidays comes in Q4. The high volume of app opening was seeing from Thursdays to Sundays in both the quarters. But in Q4 on Mondays, there is also a high volume of openings. <br>